---
title: "MetaPR2 - Chapter metabarcoding 2020"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
---

# Initialize

```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               eval = FALSE)
opts_knit$set(width=75)
```

```{r, message=FALSE, warning=FALSE}
# tidyr libraries
  library(stringr)
  library(ggplot2)
  library(dplyr)
  library(tidyr)
  library(tibble)

# to import and export data
  library(rio) 

# to format tables
  library(knitr)
  library(kableExtra)
  

# Database libraries
  library(DBI)
  library(RMySQL)
```

## Database functions

```{r}
db_info <- function(database_name,
                    file_cnf = "mysql/my.cnf")  {
# To do, try https://theautomatic.net/2019/06/25/how-to-hide-a-password-in-r-with-the-keyring-package/


  db_pr2_google <- list(dbname='pr2',
                        default.file=file_cnf,
                        groups="google-pr2")
  
  db_metapr2_google <- list(dbname='metapr2',
                        default.file=file_cnf,
                        groups="google-metapr2")
  dbinfo <- NULL

  if (database_name == "pr2_google") {db_info=db_pr2_google}
  if (database_name == "metapr2_google") {db_info=db_metapr2_google}

    return(db_info)
}

# ====================================

db_connect <- function(db_info)  {

db <- dbConnect(MySQL(),  default.file=db_info$default.file,
                          groups=db_info$groups,
                          dbname=db_info$dbname)

    return(db)
}


# ====================================

db_disconnect <- function(db_con)  {
    dbDisconnect(db_con)
    return(TRUE)
}

```

# Read the metaPR2 tables

```{r}
  db_con <- db_connect(db_info("metapr2_google"))
  datasets <- tbl(db_con, "metapr2_datasets") %>% collect()
  metadata <- tbl(db_con, "metapr2_metadata") %>% collect()
  samples <- tbl(db_con, "metapr2_samples") %>% collect()
  db_disconnect(db_con)
  
  
  metapr2 <- samples %>% 
    left_join(metadata) %>% 
    left_join(datasets)
  
```

## Filter to keep only public datasets

```{r}
 samples_18S <- metapr2 %>% 
  filter(!is.na(paper_doi),
         !str_detect(ecosystem,"agriculture|terrestrial")) %>% 
  filter(str_detect(gene,"18S"))
 
 datasets_published <- datasets %>% 
  filter(!is.na(paper_doi),
         !str_detect(ecosystem,"agriculture|terrestrial"),
         !((gene == "16S rRNA") & is.na(organelle))    # Remove 16S studies not targetting plastid
         )

 datasets_18S <- datasets_published %>% 
  filter(str_detect(gene,"18S"))
```

## Data set table for paper

```{r}
table_datasets <- datasets_published %>% 
  select(organelle, gene,gene_region,primer_specificity, dataset_id, region, ecosystem, substrate_type, bioproject_accession, paper_doi) %>% 
  arrange(organelle, gene, gene_region,  region, ecosystem, substrate_type)

excel_datasets <- datasets_published %>% 
  select(-contains(c("dada2", "contact", "removed", "remark", "metapr2", "date", "path", "run")))

```

# Figures

```{r}
theme_set(theme_bw())
```


## Map of all samples for which we have coordinates

```{r fig.height=8, fig.width=12}

asv_summary <- samples_18S %>%  
      distinct(project, dataset_id, longitude, latitude) %>% 
      mutate(project =  case_when(str_detect(project, "OSD") ~ "OSD",
                                  str_detect(project, "Tara") ~ "Tara",
                                  str_detect(project, "Malaspina") ~ "Malaspina",
                                  TRUE ~ "Other"))

n_datasets = length(unique(asv_summary$dataset_id))
    

  world <- map_data("world")

  fig_map_stations <- ggplot() + 
    geom_polygon(data = world, aes(x=long, y = lat, group = group), fill="grey") + 
    coord_fixed(1.3) +
    geom_point(data=asv_summary, aes(x=longitude, y=latitude, color = project), size=2.5) + 
    # geom_point(data=asv_summary, aes(x=longitude, y=latitude, size=n_samples), fill="blue", shape=21) + 
    # scale_size_area(trans="log10") +
    scale_color_viridis_d(option = "magma") +
    labs(title = str_c("Map fo samples for ", n_datasets," datasets.") ,
         colors = "Project",
         x = "Longitude",
         y = "Latitude") +
    scale_x_continuous(breaks = (-4:4) * 45) +
    scale_y_continuous(breaks = (-2:2) * 30)
  
  print(fig_map_stations)

```

## Number of studies per year

```{r fig.height=8, fig.width=12}

datasets_year <- datasets_published %>% 
  mutate(sequencing_technology = case_when(str_detect(sequencing_technology, "Illumina") ~ "Illumina",
                                           str_detect(sequencing_technology, "454") ~ "454",
                                            TRUE ~ "Other")) %>% 
  distinct(paper_doi, paper_year, sequencing_technology) %>% 
  filter(paper_year < 2021) %>% 
  count(paper_year, sequencing_technology) 

fig_year <- ggplot(datasets_year) +
  geom_col(aes(x=paper_year, y=n, fill=sequencing_technology), position= position_dodge2(preserve = "single", padding = 0), width=0.7) +
  scale_fill_brewer() +
  scale_x_continuous(limits=c(2008,2021), breaks = 2009:2020) +
  labs(x = "Year", y="Number of papers published", fill = "Technology") +
  dvutils::theme_dviz_hgrid()
  
  
  print(fig_year)

```

# Tables

## Genes

```{r}
table_gene <- datasets_published %>% 
  mutate(organelle = replace_na(organelle, ""), 
        gene_organelle = str_c(gene, organelle, sep = " ")) %>% 
  count(gene_organelle) %>% 
  arrange(-n) 

table_gene  
```

## Taxonomic groups

```{r}
table_taxo <- datasets_published %>% 
  filter(!is.na(primer_specificity)) %>% 
  count(primer_specificity) %>% 
  arrange(-n) 

table_taxo  
```


## Primers

### Read the primers tables from pr2

Only keep the 18S primers with V region

```{r}

# Read from local database
pr2_db <- db_info("pr2_google")
pr2_db_con <- db_connect(pr2_db)

primers <- tbl(pr2_db_con, "pr2_primers") %>% 
  collect()
primer_sets_all <- tbl(pr2_db_con, "pr2_primer_sets") %>% 
  collect()

disconnect <- db_disconnect(pr2_db_con)


  
primer_sets <- primer_sets_all %>% 
  filter(gene == "18S rRNA")
    
  

primer_sets <- primer_sets %>% 
  left_join(select(primers, 
                   primer_id, 
                   fwd_name=name,
                   fwd_seq=sequence, 
                   fwd_start_yeast= start_yeast, 
                   fwd_end_yeast= end_yeast), 
            by = c("fwd_id" = "primer_id")) %>% 
  left_join(select(primers, 
                 primer_id, 
                 rev_name=name,
                 rev_seq=sequence, 
                 rev_start_yeast= start_yeast, 
                 rev_end_yeast= end_yeast), 
          by = c("rev_id" = "primer_id")) %>% 
  mutate(length_yeast = rev_end_yeast - fwd_start_yeast + 1) %>% 
  select(gene_region, specificity, 
         primer_set_id, primer_set_name,
         contains("fwd"), 
         contains("rev"),
         length_yeast, 
         reference:remark) %>%
  select(-fwd_id, -rev_id) %>% 
  arrange(gene_region,  fwd_start_yeast, rev_start_yeast) %>% 
  mutate(specific = ifelse(is.na(specificity), "general", "specific")) %>% 
  relocate(specific, .before = specificity)
  

```


### Primers for 18S

```{r}


 n_rows = 6

 primers_usage_fwd <- datasets_18S %>% 
  count(primer_fwd_name,primer_fwd_seq, gene_region) %>% 
  arrange(-n) %>% 
  slice_head(n=n_rows) %>% 
  left_join(select(primers, sequence, direction, reference, doi), by = c("primer_fwd_seq" = "sequence")) %>% 
  rename(name = primer_fwd_name,
         sequence = primer_fwd_seq) 
   
 
primers_usage_rev <- datasets_18S %>% 
  count(primer_rev_name,primer_rev_seq, gene_region) %>% 
  arrange(-n) %>% 
  slice_head(n=n_rows) %>% 
  left_join(select(primers, sequence, direction, reference, doi), by = c("primer_rev_seq" = "sequence")) %>% 
  rename(name = primer_rev_name,
         sequence = primer_rev_seq) 

 table_primers <- bind_rows(primers_usage_fwd, primers_usage_rev) %>% 
   relocate(n, .after = last_col())

 table_primer_sets_usage <- datasets_18S %>% 
  count(primer_fwd_name,primer_rev_name, gene_region) %>% 
  arrange(-n) %>% 
  rownames_to_column() %>%  
  mutate(rowname = as.numeric(rowname))  %>% 
  mutate(primer_fwd_name = case_when(rowname > n_rows ~ "Other",
                                    TRUE ~ primer_fwd_name) ,
         # primer_fwd_seq = case_when(rowname > n_rows ~ NA_character_,
         #                            TRUE ~ primer_fwd_seq),
         primer_rev_name = case_when(rowname > n_rows ~ NA_character_,
                                    TRUE ~ primer_rev_name) ,
         # primer_rev_seq = case_when(rowname > n_rows ~ NA_character_,
         #                            TRUE ~  primer_rev_seq) , 
         gene_region  = case_when(rowname > n_rows ~ NA_character_,
                                    TRUE ~ gene_region)) %>% 
  count(primer_fwd_name,primer_rev_name, gene_region, wt= n) %>% 
  arrange(-n)  
 
 top <- table_primer_sets_usage %>% 
   filter(primer_fwd_name != "Other")
 bottom <- table_primer_sets_usage %>% 
   filter(primer_fwd_name == "Other") 
 table_primer_sets_usage <- bind_rows(top, bottom)
 
 table_primer_sets_region <- datasets_18S %>% 
  count(gene_region) %>% 
  arrange(-n) %>% 
  mutate(gene_region =   forcats::fct_lump(gene_region, n = 2 , w = n)) %>% 
  count(gene_region, wt = n)  
 
 table_primers
 
 table_primer_sets_region
 
 table_primer_sets_usage

```

## Methods of analysis

```{r}
table_method <- datasets_18S %>% 
  separate_rows(processing_pipeline_original, sep = ", ") %>% 
  mutate(method = case_when(str_detect(processing_pipeline_original, "swarm") ~ "swarm",
                            str_detect(processing_pipeline_original, "mothur") ~ "mothur",
                            str_detect(processing_pipeline_original, "dada2") ~ "dada2",
                            str_detect(processing_pipeline_original, "Qiime") ~ "QIIME",
                            str_detect(processing_pipeline_original, "[USEARCH|UCHIME|UCLUST]") ~ "USEARCH",
                            # str_detect(processing_pipeline_original, "vsearch") ~ "vsearch",
                            TRUE ~ "Other")) %>% 
  count(method) %>% 
  arrange(-n) 

table_method  
```


## Data availability
```{r}
table_availability <- datasets_published %>% 
  count(data_available) %>% 
  mutate(percentage = n/sum(n)*100)
  
table_availability
```

## Ecosystem and substrate

```{r}

table_ecosystem <- datasets_published %>%
  select(ecosystem) %>% 
  separate_rows(ecosystem, sep = ", ") %>%
  mutate(ecosystem = str_replace(ecosystem,"oceanic", "marine oceanic")) %>%
  mutate(ecosystem = str_replace(ecosystem,"coastal", "marine coastal")) %>% 
  count(ecosystem) %>%
  arrange(desc(n))

table_ecosystem


table_substrate <- datasets_published %>%
  select(substrate_type) %>% 
  separate_rows(substrate_type, sep = ", ") %>%
  count(substrate_type) %>%
  arrange(desc(n))

table_substrate
```


# Export to Excel

```{r}
rio::export(excel_datasets, "../metapr2_datasets.xlsx", zoom = 90, firstRow = TRUE, firstActiveCol = 4 )
```


# Export to Latex tables


```{r}

table_latex <- function(table, 
                        caption, 
                        label, 
                        colnames, 
                        align, 
                        digits = 2, 
                        scale_down = FALSE,
                        longtable = FALSE) {
  
  options(knitr.kable.NA = "")
  
  latex_options = c("hold_position")
  if(scale_down) latex_options = c("scale_down", latex_options)

  
  kbl( table, 
       caption = caption, 
       label=label, 
       col.names = colnames,
       align = align,
       digits = digits,
       linesep = "",
       booktabs = TRUE, 
       format = "latex", 
       escape = T,
       longtable = longtable) %>% 
  kable_styling(latex_options = latex_options,
                position="center")  

  }


```



## File path

```{r, eval=TRUE}

table_path = "../../chapter_metabarcoding_overleaf/tables/"

full_file_name <- function(file_name) {str_c(table_path, file_name)}

```

## Table - Genes and Gene Regions

```{r, eval=TRUE}

table <- table_primer_sets_region

table_1 <- table_gene %>% rename(type = gene_organelle)
table_2 <- table_primer_sets_region %>% rename(type = gene_region)
table <- bind_rows(table_1, table_2)

caption <- "Most frequently used gene markers and 18S rRNA region in eukaryotic metabarcoding studies of aquatic phytoplankton (see Table~\\ref{tab:datasets}) with the number of papers (N) where used."
label <- "gene"
file <- str_c("table_", label, ".tex")
colnames <- c( "Type", "N")
align <- c("l","c" )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, scale_down = FALSE) %>% 
  pack_rows(index=c("Gene marker" = nrow(table_1), 
                    "18S region" = nrow(table_2))) 

cat(table_out, file = full_file_name(file))

```


## Table - Primers

```{r, eval=TRUE}

table <- table_primers
caption <- "Eukaryotic 18S rRNA primers most often used in metabarcoding studies for aquatic phytoplankton and protists (see Table~\\ref{tab:datasets}) with the number of papers (N) where used."
label <- "primers"
file <- "table_primers.tex"
colnames <- c( "Name", "Sequence", "Region", "Direction", "Reference", "DOI", "N")
align <- c("l","l", "c","c","l","l","c" )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, scale_down = TRUE) 

cat(table_out, file = full_file_name(file))

```



## Table - Primer sets

```{r, eval=TRUE}

table <- table_primer_sets_usage
caption <- "18S rRNA primer sets most often used in metabarcoding studies of aquatic phytoplankton and protists (see Table~\\ref{tab:datasets}) with the number of papers (N) where used.  Refer to Table~\\ref{tab:primers} for sequence and reference of primers."
label <- "primer_sets"
file <- str_c("table_", label, ".tex")
colnames <- c( "Primer fwd", "Primer rev", "Region", "N")
align <- c("l","l", "c","c" )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, scale_down = FALSE) 

cat(table_out, file = full_file_name(file))


```
## Table - Method

```{r, eval=TRUE}

table <- table_method
caption <- "Software most often used to process sequence in metabarcoding studies of aquatic phytoplankton and protists (see Table~\\ref{tab:datasets}) with the number of papers (N) where used."
label <- "method"
file <- str_c("table_", label, ".tex")
colnames <- c( "Software", "N")
align <- c("l","c" )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, scale_down = FALSE) 

cat(table_out, file = full_file_name(file))


```

## Table - Ecosystems and substrates

```{r, eval=TRUE}

table_1 <- table_ecosystem %>% rename(type = ecosystem)
table_2 <- table_substrate %>% rename(type = substrate_type)
table <- bind_rows(table_1, table_2)

caption <- "Ecosystems and substrates targeted by aquatic metabarcoding studies (see Table~\\ref{tab:datasets})."
label <- "ecosystem"
file <- str_c("table_", label, ".tex")
colnames <- c( "Type", "N")
align <- c("l","c" )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, scale_down = FALSE)%>% 
  pack_rows(index=c("Ecosystem" = nrow(table_1), 
                    "Substrate" = nrow(table_2))) 

cat(table_out, file = full_file_name(file))

```


## Table - Vertical fluxes - Andres

```{r, eval=TRUE}

table_traps <- rio::import(full_file_name("table_traps.xlsx"), range = "A3:J13")

table <- table_traps

caption <- "Studies of vertical fluxes. na: not applicable; nd: not determined."
label <- "traps"
file <- str_c("table_", label, ".tex")
colnames <- c("Study", "Region", "Surface Chla (range, mg/m3)", "Sampling Method", "Depth of traps deployment (m)", "Duration of deployment", "Preservation", "Molecular approach", "Gene marker", "Main Finding")
align <- c(rep("p{2cm} ",9),"p{5cm}" )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, longtable = TRUE)

cat(table_out, file = full_file_name(file))


```

## Supplementary Table - Data sets 

```{r, eval=TRUE}

table <- table_datasets


caption <- "List of datasets and studies considered in this Chapter."
label <- "datasets"
file <- str_c("table_", label, ".tex")
colnames <- c("Organelle", "Gene", "Region", "Specificity", "ID", "Area", "Ecosystem", "Substrate", "Bioproject", "DOI")
align <- c(rep("c", 5), rep("p{3cm} ",5) )

table_out <- table_latex(table, caption, label, colnames, align, digits=0, longtable = TRUE)

cat(table_out, file = full_file_name(file))


```


# Export figures

```{r, eval=TRUE}

fig_path <- function (file) str_c("../../chapter_metabarcoding_overleaf/figs/", file)


ggsave(plot= fig_map_stations , filename=fig_path("fig_map_stations.pdf"), 
       width = 16 , height = 12, scale=1.8, units="cm", useDingbats=FALSE) 

ggsave(plot= fig_year , filename=fig_path("fig_year.pdf"), 
       width = 16 , height = 12, scale=1.8, units="cm", useDingbats=FALSE) 


```


