---
title: "MetaPR2 - Chapter metabarcoding 2020"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
---

# Initialize

```{r init, echo=TRUE, message=FALSE, warning=FALSE}

source("init.R")
```

# Read the metaPR2 tables

```{r}
  db_con <- db_connect(db_info("metapr2_google"))
  datasets <- tbl(db_con, "metapr2_datasets") %>% collect()
  metadata <- tbl(db_con, "metapr2_metadata") %>% collect()
  samples <- tbl(db_con, "metapr2_samples") %>% collect()
  db_disconnect(db_con)
  
  
  metapr2 <- samples %>% 
    left_join(metadata) %>% 
    left_join(datasets)
  
```

## Filter to keep only public datasets

```{r}
 samples_18S <- metapr2 %>% 
  filter(!is.na(paper_doi),
         !str_detect(ecosystem,"agriculture|terrestrial")) %>% 
  filter(str_detect(gene,"18S"))
 
 datasets_published <- datasets %>% 
  filter(!is.na(paper_doi),
         !str_detect(ecosystem,"agriculture|terrestrial"),
         !((gene == "16S rRNA") & is.na(organelle))    # Remove 16S studies not targetting plastid
         )

 datasets_18S <- datasets_published %>% 
  filter(str_detect(gene,"18S"))
```

## Data set table for paper

```{r}
table_datasets <- datasets_published %>% 
  select(organelle, gene,gene_region,primer_specificity, dataset_id, region, ecosystem, substrate_type, bioproject_accession, paper_doi) %>% 
  arrange(organelle, gene, gene_region,  region, ecosystem, substrate_type)

excel_datasets <- datasets_published %>% 
  select(-contains(c("dada2", "contact", "removed", "remark", "metapr2", "date", "path", "run")))

```

# Figures

```{r}
theme_set(theme_bw())
```


## Map of all samples for which we have coordinates

```{r fig.height=8, fig.width=12}

asv_summary <- samples_18S %>%  
      distinct(project, dataset_id, longitude, latitude) %>% 
      mutate(project =  case_when(str_detect(project, "OSD") ~ "OSD",
                                  str_detect(project, "Tara") ~ "Tara",
                                  str_detect(project, "Malaspina") ~ "Malaspina",
                                  TRUE ~ "Other")) %>% 
      mutate(project = forcats::fct_relevel(project, c("OSD", "Tara","Malaspina", "Other")))

n_datasets = length(unique(asv_summary$dataset_id))

# Using world map

  world <- map_data("world")
  
  fig_map_stations <- ggplot() + 
    geom_polygon(data = world, aes(x=long, y = lat, group = group), fill="grey") +
    coord_fixed(1.3) +
    # coord_map("mollweide") +
    geom_point(data=asv_summary, aes(x=longitude, y=latitude, color = project), size=2.5) + 
    scale_color_viridis_d(option = "magma") +
    labs(title = str_c("Map fo samples for ", n_datasets," datasets.") ,
         colors = "Project",
         x = "Longitude",
         y = "Latitude") +
    scale_x_continuous(breaks = (-4:4) * 45) +
    scale_y_continuous(breaks = (-2:2) * 30)
  
  print(fig_map_stations)
  
# Using sf
  
  # See for removing lines : https://stackoverflow.com/questions/49836184/cant-remove-gridlines-when-plotting-with-geom-sf
    

  world <- rnaturalearth::ne_countries(scale = "medium", returnclass = "sf")
  
  crs_longlat <- "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"
  crs_robinson = "+proj=robin +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m no_defs"
  
  asv_summary_sf <- asv_summary %>% 
    filter(!is.na(longitude)) %>% 
    sf::st_as_sf( coords = c("longitude", "latitude"),
                crs = crs_longlat) 
  
  # boundary <- sf::st_multilinestring(list(rbind(c(-180,90), c(180,90)), rbind(c(180,-90), c(-180,-90))))
  # boundary <- sf::st_sfc(boundary, crs=crs_longlat)
  # boundary <- sf::st_sf(geometry = boundary)

  fig_map_stations <- ggplot() + 
    geom_sf(data = world, color="grey70", fill="grey70")  +
    geom_sf(data=asv_summary_sf, aes(color=project), size=2.5) +
    # geom_sf(data=boundary) +
    theme(panel.border=element_blank(),
          panel.grid = element_line(colour = "grey70", size=2),
          axis.text.x= element_blank(),
          axis.text.y = element_blank()
          # plot.margin=margin(0,0,0,0),
          # aspect.ratio = 0.5
          ) + 
    scale_color_viridis_d(option = "viridis") +
    # scale_color_brewer() +
    labs(
         # title = str_c("Map of samples for ", n_datasets," datasets.") ,
         color = "Project",
         x = "",
         y = "") +
    scale_x_continuous(breaks = c(-180, 180)) +
    scale_y_continuous(breaks=c(-89.999, 89.999)) +
    coord_sf(crs = "+proj=robin +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m no_defs", expand = TRUE) 
  
  print(fig_map_stations)
  
```


## Number of studies per year

```{r fig.height=8, fig.width=12}

datasets_year <- datasets_published %>% 
  mutate(sequencing_technology = case_when(str_detect(sequencing_technology, "Illumina") ~ "Illumina",
                                           str_detect(sequencing_technology, "454") ~ "454",
                                            TRUE ~ "Other")) %>% 
  distinct(paper_doi, paper_year, sequencing_technology) %>% 
  filter(paper_year < 2021) %>% 
  count(paper_year, sequencing_technology) 

fig_year <- ggplot(datasets_year) +
  geom_col(aes(x=paper_year, y=n, fill=sequencing_technology), position= position_dodge2(preserve = "single", padding = 0), width=0.7) +
  scale_fill_brewer() +
  scale_x_continuous(limits=c(2008,2021), breaks = 2009:2020) +
  labs(x = "Year", y="Number of papers published", fill = "Technology") +
  dvutils::theme_dviz_hgrid()
  
  
  print(fig_year)

```

# Tables

## Genes

```{r}
table_gene <- datasets_published %>% 
  mutate(organelle = replace_na(organelle, ""), 
        gene_organelle = str_c(gene, organelle, sep = " ")) %>% 
  count(gene_organelle) %>% 
  arrange(-n) 

table_gene  
```

## Taxonomic groups

```{r}
table_taxo <- datasets_published %>% 
  filter(!is.na(primer_specificity)) %>% 
  count(primer_specificity) %>% 
  arrange(-n) 

table_taxo  
```


## Primers

### Read the primers tables from pr2

Only keep the 18S primers with V region

```{r}

# Read from local database
pr2_db <- db_info("pr2_google")
pr2_db_con <- db_connect(pr2_db)

primers <- tbl(pr2_db_con, "pr2_primers") %>% 
  collect()
primer_sets_all <- tbl(pr2_db_con, "pr2_primer_sets") %>% 
  collect()

disconnect <- db_disconnect(pr2_db_con)


  
primer_sets <- primer_sets_all %>% 
  filter(gene == "18S rRNA")
    
  

primer_sets <- primer_sets %>% 
  left_join(select(primers, 
                   primer_id, 
                   fwd_name=name,
                   fwd_seq=sequence, 
                   fwd_start_yeast= start_yeast, 
                   fwd_end_yeast= end_yeast), 
            by = c("fwd_id" = "primer_id")) %>% 
  left_join(select(primers, 
                 primer_id, 
                 rev_name=name,
                 rev_seq=sequence, 
                 rev_start_yeast= start_yeast, 
                 rev_end_yeast= end_yeast), 
          by = c("rev_id" = "primer_id")) %>% 
  mutate(length_yeast = rev_end_yeast - fwd_start_yeast + 1) %>% 
  select(gene_region, specificity, 
         primer_set_id, primer_set_name,
         contains("fwd"), 
         contains("rev"),
         length_yeast, 
         reference:remark) %>%
  select(-fwd_id, -rev_id) %>% 
  arrange(gene_region,  fwd_start_yeast, rev_start_yeast) %>% 
  mutate(specific = ifelse(is.na(specificity), "general", "specific")) %>% 
  relocate(specific, .before = specificity)
  

```


### Primers for 18S

```{r}


 n_rows = 6

 primers_usage_fwd <- datasets_18S %>% 
  count(primer_fwd_name,primer_fwd_seq, gene_region) %>% 
  arrange(-n) %>% 
  slice_head(n=n_rows) %>% 
  left_join(select(primers, sequence, direction, reference, doi), by = c("primer_fwd_seq" = "sequence")) %>% 
  rename(name = primer_fwd_name,
         sequence = primer_fwd_seq) 
   
 
primers_usage_rev <- datasets_18S %>% 
  count(primer_rev_name,primer_rev_seq, gene_region) %>% 
  arrange(-n) %>% 
  slice_head(n=n_rows) %>% 
  left_join(select(primers, sequence, direction, reference, doi), by = c("primer_rev_seq" = "sequence")) %>% 
  rename(name = primer_rev_name,
         sequence = primer_rev_seq) 

 table_primers <- bind_rows(primers_usage_fwd, primers_usage_rev) %>% 
   relocate(n, .after = last_col())

 table_primer_sets_usage <- datasets_18S %>% 
  count(primer_fwd_name,primer_rev_name, gene_region) %>% 
  arrange(-n) %>% 
  rownames_to_column() %>%  
  mutate(rowname = as.numeric(rowname))  %>% 
  mutate(primer_fwd_name = case_when(rowname > n_rows ~ "Other",
                                    TRUE ~ primer_fwd_name) ,
         # primer_fwd_seq = case_when(rowname > n_rows ~ NA_character_,
         #                            TRUE ~ primer_fwd_seq),
         primer_rev_name = case_when(rowname > n_rows ~ NA_character_,
                                    TRUE ~ primer_rev_name) ,
         # primer_rev_seq = case_when(rowname > n_rows ~ NA_character_,
         #                            TRUE ~  primer_rev_seq) , 
         gene_region  = case_when(rowname > n_rows ~ NA_character_,
                                    TRUE ~ gene_region)) %>% 
  count(primer_fwd_name,primer_rev_name, gene_region, wt= n) %>% 
  arrange(-n)  
 
 top <- table_primer_sets_usage %>% 
   filter(primer_fwd_name != "Other")
 bottom <- table_primer_sets_usage %>% 
   filter(primer_fwd_name == "Other") 
 table_primer_sets_usage <- bind_rows(top, bottom)
 
 table_primer_sets_region <- datasets_18S %>% 
  count(gene_region) %>% 
  arrange(-n) %>% 
  mutate(gene_region =   forcats::fct_lump(gene_region, n = 2 , w = n)) %>% 
  count(gene_region, wt = n)  
 
 table_primers
 
 table_primer_sets_region
 
 table_primer_sets_usage

```

## Methods of analysis

```{r}
table_method <- datasets_18S %>% 
  separate_rows(processing_pipeline_original, sep = ", ") %>% 
  mutate(method = case_when(str_detect(processing_pipeline_original, "swarm") ~ "swarm",
                            str_detect(processing_pipeline_original, "mothur") ~ "mothur",
                            str_detect(processing_pipeline_original, "dada2") ~ "dada2",
                            str_detect(processing_pipeline_original, "Qiime") ~ "QIIME",
                            str_detect(processing_pipeline_original, "[USEARCH|UCHIME|UCLUST]") ~ "USEARCH",
                            # str_detect(processing_pipeline_original, "vsearch") ~ "vsearch",
                            TRUE ~ "Other")) %>% 
  count(method) %>% 
  arrange(-n) 

table_method  
```


## Data availability
```{r}
table_availability <- datasets_published %>% 
  count(data_available) %>% 
  mutate(percentage = n/sum(n)*100)
  
table_availability
```

## Ecosystem and substrate

```{r}

table_ecosystem <- datasets_published %>%
  select(ecosystem) %>% 
  separate_rows(ecosystem, sep = ", ") %>%
  mutate(ecosystem = str_replace(ecosystem,"oceanic", "marine oceanic")) %>%
  mutate(ecosystem = str_replace(ecosystem,"coastal", "marine coastal")) %>% 
  count(ecosystem) %>%
  arrange(desc(n))

table_ecosystem


table_substrate <- datasets_published %>%
  select(substrate_type) %>% 
  separate_rows(substrate_type, sep = ", ") %>%
  count(substrate_type) %>%
  arrange(desc(n))

table_substrate
```


# Export to Excel

```{r}
rio::export(excel_datasets, "../metapr2_datasets.xlsx", zoom = 90, firstRow = TRUE, firstActiveCol = 4 )
```


